File: packages/capacitor-notificare-push/android/src/main/java/re/notifica/push/capacitor/EventBroker.kt
package re.notifica.push.capacitor

import com.getcapacitor.JSObject
import org.json.JSONObject

internal typealias NotifyListenersFunction = (name: String, data: JSObject?, retainUntilConsumed: Boolean) -> Unit

internal object EventBroker {
    private var notifyListenersFunction: NotifyListenersFunction? = null
    private val eventQueue = mutableListOf<Event>()

    fun setup(notifyListenersFunction: NotifyListenersFunction) {
        this.notifyListenersFunction = notifyListenersFunction

        if (eventQueue.isNotEmpty()) {
            logger.debug("Processing event queue with ${eventQueue.size} items.")
            eventQueue.forEach { notifyListenersFunction(it.name, it.data, it.retainUntilConsumed) }
            eventQueue.clear()
        }
    }

    fun dispatchEvent(name: String, data: JSONObject?, retainUntilConsumed: Boolean = false) {
        dispatchEvent(name, data?.let { JSObject.fromJSONObject(it) }, retainUntilConsumed)
    }

    fun dispatchEvent(name: String, data: JSObject?, retainUntilConsumed: Boolean = false) {
        val notifyListenersFunction = notifyListenersFunction ?: run {
            eventQueue.add(Event(name, data, retainUntilConsumed))
            return
        }

        notifyListenersFunction(name, data, retainUntilConsumed)
    }

    private data class Event(
        val name: String,
        val data: JSObject?,
        val retainUntilConsumed: Boolean
    )
}

File: packages/capacitor-notificare/ios/Plugin/EventBroker.swift
import Foundation
import Capacitor

internal typealias NotifyListenersFunction = (_ eventName: String, _ data: [String: Any]?, _ retainUntilConsumed: Bool) -> Void

internal class EventBroker {
    private init() {}
    
    internal static let instance: EventBroker = .init()
    
    private var notifyListenersFunction: NotifyListenersFunction?
    private var queue:  [Event] = []
    
    func setup(_ notifyListenersFunction: @escaping NotifyListenersFunction) {
        self.notifyListenersFunction = notifyListenersFunction
        
        if !queue.isEmpty {
            logger.debug("Processing event queue with ${eventQueue.size} items.")
            queue.forEach { dispatchEvent($0.name, data: $0.data)}
            queue.removeAll()
        }
    }
    
    internal func dispatchEvent(_ name: String, data: [String: Any]?, retainUntilConsumed: Bool = false) {
        guard let notifyListenersFunction = notifyListenersFunction else {
            queue.append(Event(name: name, data: data, retainUntilConsumed: retainUntilConsumed))
            return
        }
        
        notifyListenersFunction(name, data, retainUntilConsumed)
    }
    
    internal struct Event {
        let name: String
        let data: [String: Any]?
        let retainUntilConsumed: Bool
    }
}
